### PR TITLE
Feature/develop2 profiles api

### PR DESCRIPTION
**Pull request by memsharded**
New ``profiles`` API

### CODE REVIEW

File: conans/cli/commands/profile.py [hunk 1]
```diff
@@ -0,0 +1,131 @@
+from conans.cli.command import conan_command, conan_subcommand, Extender, COMMAND_GROUPS
+from conans.cli.commands import json_formatter
+from conans.cli.output import ConanOutput
+
+
+def profiles_cli_output(profiles):
+    host, build = profiles
+    output = ConanOutput()
```
Comment (czoido) [inline conans/cli/commands/profile.py]: Shouldn't this be `cli_out_write` ? I think `ConanOutput` is writting to the **stderr**, not **stdout**.

**Comment by czoido** [comment]
It's not critical, but maybe I would change the UI for `conan profile show`, make the profile positional and add the context as an argument (with default value `host` maybe)?
I find a little bit strange doing something like `conan profile show -pr:h=default`

File: conans/cli/commands/profile.py (hunk 1 above)
Comment (memsharded) [reply conans/cli/commands/profile.py]: Yes, you are right, lets fix it.

**Comment by memsharded** [comment]
>  I find a little bit strange doing something like conan profile show -pr:h=default

The rational of this is to show the profiles that will result from some command line arguments, including something like ``conan profile show -pr=default -s build_type=Debug`` for example. Using exactly the same command line arguments that are passed to other commands.

I also thought about splitting, doing 1 profile either build or host (it is relevant too, might need some arg like ``--build`` to specify), at a time, but the UI seemed way more ad-hoc for this particular command, so I didn't like it much. Lets review it together.

**Review by czoido — APPROVED** [review]

File: conans/cli/api/subapi/profiles.py [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+from conans.model.profile import Profile
+from conans.util.files import save
+
+
+class ProfilesAPI:
+
+    def __init__(self, conan_api):
+        self.conan_api = conan_api
+
+    def get_profile(self, profiles=None, settings=None, options=None, conf=None,
```
Comment (lasote) [inline conans/cli/api/subapi/profiles.py]:
```suggestion
    def combine(self, profiles=None, settings=None, options=None, conf=None,
```

File: conans/cli/api/subapi/profiles.py [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+                    cwd=None, build_profile=False):
+        cache = ClientCache(self.conan_api.cache_folder)
+        loader = ProfileLoader(cache)
+        env = None  # TODO: Not handling environment
+        profile = loader.from_cli_args(profiles, settings, options, env, conf, cwd, build_profile)
+        return profile
+
+    def profile_path(self, profile, cwd=None):
```
Comment (lasote) [inline conans/cli/api/subapi/profiles.py]:
```suggestion
    def get_path(self, profile, cwd=None):
```

File: conans/cli/api/subapi/profiles.py [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+            for current_directory, _, files in os.walk(profiles_path, followlinks=True):
+                for filename in files:
+                    rel_path = os.path.relpath(os.path.join(current_directory, filename),
+                                               profiles_path)
+                    profiles.append(rel_path)
+
+        profiles.sort()
+        return profiles
+
+    def detect_profile(self, profile_name=None, force=False):
```
Comment (lasote) [inline conans/cli/api/subapi/profiles.py]:
```suggestion
    def detect(self, profile_name=None, force=False):
```

File: conans/cli/api/subapi/profiles.py (hunk 2 above)
Comment (lasote) [reply conans/cli/api/subapi/profiles.py]: I don't like the API having the `profiles` with `None` and then having to say, hey, but this is supposed to be a `build` profile, so read me the build default... too much, road to messy API again.
I would prefer to have a `def get(self, profile_name)`, even a `get_build_default()` and `get_host_default()` and pass explicit argument to a `def combine(self, profiles, settings=None, options=None....)`.
That would probably require to refactor the `loader.from_cli_args` to extract the `calculation` of the default profiles.

Comment (lasote) [reply conans/cli/api/subapi/profiles.py]: If the `Profile` object is managed in the API (I totally agree) I would move it to the `model.py` (or create a python package if better)

File: conans/cli/api/subapi/profiles.py (hunk 4 above)
Comment (lasote) [reply conans/cli/api/subapi/profiles.py]: I would rather have a `def detect(self) -> Profile` method and a `def save(profile, file_name, force=False)`

**Review by lasote — CHANGES_REQUESTED** [review]

File: conans/cli/api/subapi/profiles.py (hunk 2 above)
Comment (memsharded) [reply conans/cli/api/subapi/profiles.py]: Agree with the first, reviewed, lets see if now is better.
The second one must wait, current ``Profile`` is too dirty, that requires more work to do, too much here.

File: conans/cli/api/subapi/profiles.py [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+        it loads the "profiles", composing them in order (last profile has priority), and finally
+        adding the individual settings, options (priority over the profiles)
+        """
+        assert profiles and isinstance(profiles, list), "Please provide at least 1 profile"
+        loader = ProfileLoader(self._cache)
+        env = None  # TODO: Not handling environment
+        profile = loader.from_cli_args(profiles, settings, options, env, conf, cwd)
+        return profile
+
+    def get_path(self, profile, cwd=None, exists=True):
```
Comment (lasote) [inline conans/cli/api/subapi/profiles.py:41]: `profile` or `profile_name`?

File: conans/cli/api/subapi/profiles.py [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+
+    def get_default_build(self):
+        """
+        @return: the path to the default "build" profile, either in the cache or as defined by
+        the user in configuration
+        """
+        loader = ProfileLoader(self._cache)
+        return loader.get_default_build()
+
+    def get_profile(self, profiles, settings=None, options=None, conf=None, cwd=None):
```
Comment (lasote) [inline conans/cli/api/subapi/profiles.py:30]: `profiles` or `profiles_paths`?

File: conans/cli/api/subapi/profiles.py [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+from conans.client.profile_loader import ProfileLoader
+from conans.model.profile import Profile
+
+
+class ProfilesAPI:
+
+    def __init__(self, conan_api):
+        self._cache = ClientCache(conan_api.cache_folder)
+
+    def get_default_host(self):
```
Comment (lasote) [inline conans/cli/api/subapi/profiles.py:14]:
```suggestion
    def get_default_host_path(self):
```
?

File: conans/cli/api/subapi/profiles.py [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+        """
+        @return: the path to the default "host" profile, either in the cache or as defined by
+        the user in configuration
+        """
+        loader = ProfileLoader(self._cache)
+        return loader.get_default_host()
+
+    def get_default_build(self):
```
Comment (lasote) [inline conans/cli/api/subapi/profiles.py:22]:
```suggestion
    def get_default_build_path(self):
```
?

**Review by lasote — APPROVED** [review]

**Pull request merged**